tcmur_cmd_handler.c

```diff
@@ -181,6 +181,340 @@ struct write_same {
 	size_t iov_len;
 };
 
+struct unmap_state {
+	pthread_mutex_t lock;
+	unsigned int refcount;
+	bool error;
+};
+
+struct unmap_descriptor {
+	uint64_t offset;
+	uint64_t length;
+
+	struct tcmulib_cmd *origcmd;
+};
+
+static void handle_unmap_cbk(struct tcmu_device *dev, struct tcmulib_cmd *ucmd, int ret)
+{
+	struct unmap_descriptor *desc = ucmd->cmdstate;
+	struct tcmulib_cmd *origcmd = desc->origcmd;
+	struct unmap_state *state = origcmd->cmdstate;
+	bool error;
+
+	free(desc);
+	free(ucmd);
+
+	pthread_mutex_lock(&state->lock);
+	if (--state->refcount > 0) {
+		pthread_mutex_unlock(&state->lock);
+		return;
+	}
+	error = state->error;
```

Comment:
Oh yeah, for this case where a handler->unmap call may fail and/or a handle_split_align_unmap call may fail or both might, I think as long as one of the sense or status codes is returned it will be ok for now. We should probably prioritize them, but that looks like it will be a mess. In the next release I want to move the sense handling to runner, and then we can handle this.